File: 5_Kyu_Kata/convert-a-hex-string-to-rgb/Python/convert_a_hex_string_to_rgb_test.py
# [object Object] - Convert A Hex String To RGB  [ ID: 5282b48bb70058e4c4000fa7  (convert-a-hex-string-to-rgb) ]
# URL: https://www.codewars.com/kata/5282b48bb70058e4c4000fa7
# Category: ALGORITHMS  |  Tags: PARSING | STRINGS | ALGORITHMS
# ******************************************************************************
@test.describe('Example Tests')
def example_tests():
    test.assert_equals(hex_string_to_RGB("#FF9933"), {'r':255, 'g':153, 'b':51})
    test.assert_equals(hex_string_to_RGB("#beaded"), {'r':190, 'g':173, 'b':237})
    test.assert_equals(hex_string_to_RGB("#000000"), {'r':0, 'g':0, 'b':0})
    test.assert_equals(hex_string_to_RGB("#111111"), {'r':17, 'g':17, 'b':17})
    test.assert_equals(hex_string_to_RGB("#Fa3456"), {'r':250, 'g':52, 'b':86})

from random import randint, random

@test.describe('Random Tests')
def random_tests():

    def generate_random_case():
        res = f"#{randint(0, 0xFFFFFF):06x}"
        return res.lower() if random()<0.5 else res.upper()

    def _hex_string_to_RGB_123(hex_string): 
        return {k: int(hex_string[i:i+2], 16) for k, i in zip('rgb', range(1, 6, 2))}

    def _do_one_test():
        hex_string = generate_random_case()
        expected = _hex_string_to_RGB_123(hex_string)
        actual = hex_string_to_RGB(hex_string)
        test.assert_equals(actual, expected)

    @test.it('Random Test Cases')
    def random_test_cases():
        for _ in range(100):
            _do_one_test()

File: 5_Kyu_Kata/convert-a-hex-string-to-rgb/Swift/convertAHexStringToRgb.Test.swift
// [object Object] - Convert A Hex String To RGB  [ ID: 5282b48bb70058e4c4000fa7  (convert-a-hex-string-to-rgb) ]
// URL: https://www.codewars.com/kata/5282b48bb70058e4c4000fa7
// Category: ALGORITHMS  |  Tags: PARSING | STRINGS | ALGORITHMS
// *****************************************************************************
import XCTest
// XCTest Spec Example:
// TODO: replace with your own tests (TDD), these are just how-to examples to get you started

class SolutionTest: XCTestCase {
    static var allTests = [
        ("testHexStringToRGB", testHexStringToRGB),
        ("testRandomHexStrings", testRandomHexStrings),
    ]

    func testHexStringToRGB() {
        XCTAssertEqual(hexStringToRGB("#FF9933"), RGB(255, 153, 51))
        XCTAssertEqual(hexStringToRGB("#beaded"), RGB(190, 173, 237))
        XCTAssertEqual(hexStringToRGB("#000000"), RGB(0, 0, 0))
        XCTAssertEqual(hexStringToRGB("#111111"), RGB(17, 17, 17))
        XCTAssertEqual(hexStringToRGB("#Fa3456"), RGB(250, 52, 86))
    }
  
    func reference(_ str:String) -> RGB {
      let nums = Array(str.dropFirst())
      let r = Int(String(nums[...1]), radix: 16)!
      let g = Int(String(nums[2...3]), radix: 16)!
      let b = Int(String(nums[4...5]), radix: 16)!
      return RGB(r, g, b)
    }
  
    func getRandomString() -> String {
      let letters = "0123456789abcdefABCDEF"
      var output = "#"
      for _ in 1...6
      {
        output += String(Array(letters)[Int.random(in: 0..<22)])
      }
      return output
    }
  
    func testRandomHexStrings() {
        for _ in 1...100
        {
          let s = getRandomString()
          XCTAssertEqual(hexStringToRGB(s), reference(s))
        }
        
    }
  
}

XCTMain([
    testCase(SolutionTest.allTests)
])
